wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs

```diff
@@ -0,0 +1,239 @@
+                     logMessage)
+
+import           Cardano.Wallet.Kernel.DB.HdWallet (eskToHdRootId)
+import           Cardano.Wallet.Kernel.Types (WalletId (..))
+
+-- Internal storage necessary to smooth out the legacy 'UserSecret' API.
+data InternalStorage =
+      StorageInitialised !UserSecret
+    | StorageReleased
+
+data Keystore = Keystore (MVar InternalStorage)
+
+data KeystoreOperation =
+      KeystoreLookup
+    | KeystoreDelete
+    | KeystoreInsert
+    | KeystoreToList
+    | KeystoreRelease
+    deriving Show
+
+data KeystoreError =
+    KeystoreErrorAlreadyReleased KeystoreOperation
+    -- ^ The keystore was already released by the time the requested
+    -- operation was attempted.
+    deriving Show
+
+instance Exception KeystoreError
+
+-- | Internal monad used to smooth out the 'WithLogger' dependency imposed
+-- by 'Pos.Util.UserSecret', to not commit to any way of logging things just yet.
+newtype KeystoreM a = KeystoreM { fromKeystore :: IdentityT IO a }
+                    deriving (Functor, Applicative, Monad, MonadIO)
+
+instance HasLoggerName KeystoreM where
+    askLoggerName = return (LoggerName "Keystore")
+    modifyLoggerName _ action = action
+
+instance CanLog KeystoreM where
+    dispatchMessage _ln sev txt = logMessage sev txt
+
+{-------------------------------------------------------------------------------
+  Creating a keystore
+-------------------------------------------------------------------------------}
+
+-- | Creates a 'Keystore' using a 'bracket' pattern, where the
+-- initalisation and teardown of the resource are wrapped in 'bracket'.
+bracketKeystore :: FilePath -> (Keystore -> IO a) -> IO a
+bracketKeystore fp withKeystore =
+    bracket (newKeystore fp) releaseKeystore withKeystore
+
+-- | Creates a new keystore.
+newKeystore :: FilePath -> IO Keystore
+newKeystore fp = runIdentityT $ fromKeystore $ do
+    us <- takeUserSecret fp
+    Keystore <$> newMVar (StorageInitialised us)
+
+-- | Creates a legacy 'Keystore' by reading the 'UserSecret' from a 'NodeContext'.
+-- Hopefully this function will go in the near future.
+newLegacyKeystore :: UserSecret -> IO Keystore
+newLegacyKeystore us = Keystore <$> newMVar (StorageInitialised us)
+
+-- | Creates a legacy 'Keystore' using a 'bracket' pattern, where the
+-- initalisation and teardown of the resource are wrapped in 'bracket'.
+bracketLegacyKeystore :: UserSecret -> (Keystore -> IO a) -> IO a
+bracketLegacyKeystore us withKeystore =
+    bracket (newLegacyKeystore us)
+            (\_ -> return ()) -- Leave teardown to the legacy wallet
+            withKeystore
+
+-- | Creates a 'Keystore' out of a randomly generated temporary file (i.e.
+-- inside your $TMPDIR of choice). Suitable for testing.
+-- We don't offer a 'bracket' style here as the teardown is irrelevant, as
+-- the file is disposed automatically from being created into the
+-- OS' temporary directory.
```

Comment:
Let's rename this to `newTestKeystore`, and bang in a comment that this function has some theoretical race conditions with respect to filenames and should only be used in tests.